File: src/main/java/com/hileco/cortex/primitives/heavy/BackedVmSet.java
package com.hileco.cortex.primitives.heavy;


import com.hileco.vm.primitives.VmSet;

import java.util.HashSet;
import java.util.Set;

public class BackedVmSet<T> implements VmSet<T, BackedVmSet<T>> {
    private Set<T> backingSet;

    public BackedVmSet() {
        this.backingSet = new HashSet<>();
    }

    private BackedVmSet(Set<T> backingSet) {
        this.backingSet = new HashSet<>(backingSet);
    }

    public void close() {
    }

    public BackedVmSet<T> copy() {
        return new BackedVmSet<>(backingSet);
    }

    public boolean contains(T value) {
        return backingSet.contains(value);
    }

    public void add(T value) {
        backingSet.add(value);
    }

    public void remove(T value) {
        backingSet.remove(value);
    }

    public int size() {
        return backingSet.size();
    }
}

File: cortex-collections/src/main/kotlin/com/hileco/cortex/collections/backed/BackedVmSet.kt
package com.hileco.cortex.collections.backed


import com.hileco.cortex.collections.VmSet
import java.util.*

class BackedVmSet<T> : VmSet<T> {
    private var backingSet: MutableSet<T>

    constructor() {
        this.backingSet = HashSet()
    }

    private constructor(backingSet: Set<T>) {
        this.backingSet = HashSet(backingSet)
    }

    override fun close() {}

    override fun copy(): BackedVmSet<T> {
        return BackedVmSet(backingSet)
    }

    override fun contains(value: T): Boolean {
        return backingSet.contains(value)
    }

    override fun add(value: T) {
        backingSet.add(value)
    }

    override fun remove(value: T) {
        backingSet.remove(value)
    }

    override fun size(): Int {
        return backingSet.size
    }
}